docs/migrating-from-v3/README.md

```diff
@@ -0,0 +1,299 @@
+  - `src/store.js` moved to `src/store/index.js`;
+  - `src/router.js` renamed to `src/router/index.js`;
+- The `router` & `routerHistoryMode` options in `preset.json` is still supported for compatibility reason. But it's now recommended to use `plugins: { '@vue/cli-plugin-router': { historyMode: true } }` for better consistency.
+- `api.hasPlugin('vue-router')` is no longer supported. It's now `api.hasPlugin('router')`.
+
+### `@vue/cli-service`
+
+#### Whitespace handling in the template block
+
+To get a smaller bundle, we've disabled the `preserveWhitespace` option of `vue-template-compiler` by default in Vue CLI v3.
+
+This set comes with some caveats, however.
+
+Luckily, since the Vue 2.6 release, we can now have finer control over the whitespace handling, with the [new `whitespace` option](https://github.com/vuejs/vue/issues/9208#issuecomment-450012518). So we decided to switch over to use this new option by default in Vue CLI v4.
+
+Take the following template as an example:
+
+```html
+<p>
+  Welcome to <b>Vue.js</b> <i>world</i>.
+  Have fun!
+</p>
+```
+
+With `preserveWhitespace: false`, all whitespaces between tags are removed, so it was compiled as:
+
+```html
+<p> Welcome to <b>Vue.js</b><i>world</i>. Have fun! </p>
+```
+
+With `whitespace: 'condense'`, it is now compiled as:
+
+```html
+<p> Welcome to <b>Vue.js</b> <i>world</i>. Have fun! </p>
+```
+
+Note the **inline** whitespace between tags is now preserved.
+
+#### `vue-cli-service build --mode development`
+
+In the past, when running the `build` command in the `development` mode, the `dist` folder layout would be different from the `production` mode. Now with the following two changes, the directory structures across all modes would be the same (file names are still different - no hashes in `development` mode):
+
+- [#4323](https://github.com/vuejs/vue-cli/pull/4323) ensure consistent directory structure for all modes
+- [#4302](https://github.com/vuejs/vue-cli/pull/4302) move dev configs into serve command
+
+#### For SASS/SCSS Users
+
+Previously in Vue CLI v3, we shipped with `sass-loader@7` by default.
+
+Recently `sass-loader@8` has been out and has changed its configuration format quite a lot. Here's the release notes: <https://github.com/webpack-contrib/sass-loader/releases/tag/v8.0.0>
+
+`@vue/cli-service` will continue to support `sass-loader@7` in v4, but we strongly recommend you to take a look at the v8 release and upgrade to the latest version.
+
+#### For Less Users
+
+`less-loader` v4 is incompatible with `less` >= v3.10, see <https://github.com/less/less.js/issues/3414>.
+It's strongly recommended to upgrade to `less-loader@5` if your project depends on it.
+
+#### For CSS Module Users
+
+- [Deprecate `css.modules` in favor of `css.requireModuleExtension`](https://github.com/vuejs/vue-cli/pull/4387). This is because we've upgraded to css-loader v3 and the config format has been changed. For a more detailed explanation please follow the link.
```

Comment:
```suggestion
- [Deprecate `css.modules` in favor of `css.requireModuleExtension`](https://github.com/vuejs/vue-cli/pull/4387). This is because we've upgraded to `css-loader` v3 and the config format has been changed. For a more detailed explanation please follow the link.
```